### PR TITLE
fix(aws_ec2_metadata transform): Treat empty namespaces as not set

### PR DESCRIPTION
**Pull request by LucioFranco**
Reference #2477 

This PR is broken up into two main commits, the first commit upgrades the sink to use our internal HTTP client and the second commit allows us to treat `namespace = ""` as if it were never set.

cc @binarylogic

### CODE REVIEW

**Comment by binarylogic** [comment]
Thanks. To be clear, setting `namespace` to `""` means that all keys will be added to the root? Do you think that warrants a behavior test?

**Comment by binarylogic** [comment]
Also, I'm curious if this is problem elsewhere in Vector. I addressed that in the Schema spec [here](https://github.com/timberio/vector/pull/2428/files#diff-d70c9071a045922fceb3a69e72920175R157).

**Comment by LucioFranco** [comment]
> To be clear, setting namespace to "" means that all keys will be added to the root?

Right, this is what I am thinking because from the config in that issue it just looks like the user copied the example config. In the example it shows what the default is but since toml doesn't have optional types like Rust the only way to have the default is to not even have the config key in there. So I see this as a work around.

> Do you think that warrants a behavior test?

I added a test that should check this behavior but docker is being funky right now so I can't run the test so we will see if CI passes. (Working on fixing docker now).

I have not checked what our event code does for inserting the event since it has changed a lot recently. I would assume `".field" gets inserted at the root but I am not totally sure! Beyond that I don't see any other issues with the log nor the transform when I looked at it this morning.

**Comment by binarylogic** [comment]
I don't completely understand your answers:

> Right, this is what I am thinking

So the ec2 metadata fields are being inserted in the root of the document?

> I added a test that should check this behavior

I'm talking about `tests/behavior`, specifically.

**Comment by LucioFranco** [comment]
> So the ec2 metadata fields are being inserted in the root of the document?

No, if you have the namespace field set in the config we prepend the keys with it no matter what. Even if you have it set to "". From my initial tests with our insert code it looks like `".some_key"` actually gets discarded. I will open an issue for this with a failing test.

> I'm talking about tests/behavior, specifically.

Ah probably, I already had a similar test that I copied so it was much easier. We might want to add some behavior tests in the future though!

**Comment by binarylogic** [comment]
I still don't understand what you're saying 😆 .

> Ah probably, I already had a similar test that I copied so it was much easier. We might want to add some behavior tests in the future though!

I'm saying this is where the test belongs because it takes into account the full path. All the way from config, to parsing the config, to the expected output. Ex: what if we decide to translate `""` to `None` during config parsing as some point in the future?

**Comment by LucioFranco** [comment]
> what if we decide to translate "" to None during config parsing at some point in the future?

This is what this PR does.

> I'm saying this is where the test belongs because it takes into account the full path. All the way from config, to parsing the config, to the expected output.

The test basically just sets the config manually which should replicate a specific toml config. I will try and add a behavior test.

**Comment by LucioFranco** [comment]
Realizing we can't use the ec2 metadata transform here because there is no way to let it fetch the first events then run it through the test pipeline? In my tests within the code I have a sleep to let it make the http requests but with the behavior tests I can't replicate that.

File: src/transforms/aws_ec2_metadata.rs [hunk 1]
```diff
@@ -112,7 +112,18 @@ impl TransformConfig for Ec2Metadata {
     fn build(&self, cx: TransformContext) -> crate::Result<Box<dyn Transform>> {
         let (read, write) = evmap::new();
 
-        let keys = Keys::new(&self.namespace);
+        // Check if the namespace is set to `""` which should mean that we do
+        // not want a prefixed namespace.
+        let namespace = self.namespace.clone().and_then(|namespace| {
+            if namespace == "" {
```
Comment (lukesteensen) [inline src/transforms/aws_ec2_metadata.rs:118]: Is it worth trimming the input here to be extra defensive? i.e. should `"  "` also be treated as empty?

Comment (LucioFranco) [reply src/transforms/aws_ec2_metadata.rs:118]: I guess but at this point this just sounds like something our insert should handle?

Comment (lukesteensen) [reply src/transforms/aws_ec2_metadata.rs:118]: That is a fair point.

**Pull request merged**